src/Microsoft.ML.FastTree/TreeEnsemble/RegressionTree.cs

```diff
@@ -14,25 +14,73 @@
 using Microsoft.ML.Model;
 using Microsoft.ML.Model.Pfa;
 using Newtonsoft.Json.Linq;
-using Float = System.Single;
 
 namespace Microsoft.ML.Trainers.FastTree.Internal
 {
-    public class RegressionTree
+    public class RegressionTreeView
     {
-        private double _maxOutput;
+        /// <summary>
+        /// <see cref="RegressionTreeView"/> is an immutable wrapper over <see cref="_tree"/> for exposing some tree's attribute to users.
+        /// </summary>
+        private readonly RegressionTree _tree;
+
+        // Immutable accessors to the underlying tree structure. They help users to inspect tree learned using, for example,
+        // gradient-boosting decision tree and random forest.
+        public ReadOnlySpan<int> LteChild => _tree.LteChild;
```

Comment:
>_tree.LteChild [](start = 45, length = 14)

Could we check something about this? One thing I am concerned about is that even if a tree doesn't grow to "full size" (e.g., because of no candidate splits), the underlying tree used in training still has the full number of potential sizes allocated. So you ask for "I want 20 leaves," this will still be of size 19, even if ultimately only 10 splits are performed. One way to write a test for this "underbuilt" scenario is to use a small dataset (e.g., breast-cancer, which has under 700 leaves), use 100 for min docs per leaf, but ask for 10 leaves. (Which is impossible.) So I expect that it will produce fewer than that number of leaves.

The reason I believe this is if I view this code:

[https://github.com/dotnet/machinelearning/blob/8cdfa613c363b45de4b61abef8790943243ba818/src/Microsoft.ML.FastTree/TreeEnsemble/RegressionTree.cs#L82-L96](https://github.com/dotnet/machinelearning/blob/8cdfa613c363b45de4b61abef8790943243ba818/src/Microsoft.ML.FastTree/TreeEnsemble/RegressionTree.cs#L82-L96)

I see these arrays are allocated to the max size that *might* be needed, but I do not see that it is ever downgraded. So you might be presenting "working memory."

Reply:
Everything wrapper on pre-allocated array with size [maxLeaves -1] will be created via `new ReadOnlySpan<...>(..., 0, _tree.NumNodes)`; for example,
```csharp
        public ReadOnlySpan<int> LteChild => new ReadOnlySpan<int>(_tree.LteChild, 0, _tree.NumNodes); 
        public ReadOnlySpan<int> GtChild => new ReadOnlySpan<int>(_tree.GtChild, 0, _tree.NumNodes); 
        public ReadOnlySpan<int> NumericalSplitFeatureIndexes => new ReadOnlySpan<int>(_tree.SplitFeatures, 0, _tree.NumNodes); 
        public ReadOnlySpan<float> NumericalSplitThresholds => new ReadOnlySpan<float>(_tree.RawThresholds, 0, _tree.NumNodes);
```

---
In reply to: [251182399](https://github.com/dotnet/machinelearning/pull/2243#discussion_r251182399) [](ancestors = 251182399)